### PR TITLE
Added example demonstrating sf::VertexBuffer, sf::Shader and sf::Thread usage

### PR DESCRIPTION
**Pull request by binary1248**
See title.

### CODE REVIEW

File: examples/island/Island.cpp [hunk 1]
```diff
@@ -0,0 +1,540 @@
...
+///
+////////////////////////////////////////////////////////////
+void generateTerrain(sf::VertexBuffer& vertexBuffer)
+{
+    for (;;)
+    {
+        {
+            sf::Lock lock(workQueueMutex);
+
+            if(workQueue.empty())
```
Comment (mantognini) [inline examples/island/Island.cpp]: Just nitpicking, space missing after `if`.

File: examples/island/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,10 @@
+
+set(SRCROOT ${PROJECT_SOURCE_DIR}/examples/island)
+
+# all source files
+set(SRC ${SRCROOT}/Island.cpp)
+
+# define the island target
+sfml_add_example(island GUI_APP
+                 SOURCES ${SRC}
+                 DEPENDS sfml-graphics sfml-window sfml-system)
```
Comment (mantognini) [inline examples/island/CMakeLists.txt]: Missing `RESOURCES_DIR resources` as per recent update of `sfml_add_example`.

**Comment by binary1248** [comment]
Fixed. Can't reproduce the SIGSEGV though... Tried building with GCC and clang in all configurations I could think of. Always ran as expected.

**Comment by mantognini** [comment]
I couldn't reproduce the crash, but it wasn't systematic last time.

Here are a couple of screenshots illustrating rendering bugs. They are all captures of different runs of the example.

I played a bit with the blockCount size: it influences the size of the black strips.
broken_island_2

I also got this one, but I'm not sure how to interpret that...
broken_island_3

In the following two pictures, you can see that the bottom of the island is "broken": part of the rendering is from the previously generated island.
broken_island_1
broken_island_4

I also noticed that when I change a setting, the previous island get rendered for a very short amount of time, then the new one is displayed.

Hopefully, you'll be able to interpret those symptoms better than I.

**Comment by binary1248** [comment]
We already saw from the RenderTexture issue that MacOS currently has its issues with multi-context OpenGL usage. It wouldn't surprise me that the internal synchronization of the data transfers is somehow also broken on MacOS.

**Comment by mantognini** [comment]
Do you think it would be possible to expose some kind of sync mechanism (assuming it would fix the issue)? Maybe something special for macOS...

**Comment by binary1248** [comment]
@mantognini Give the latest version a try. I added in some synchronization between buffer updating and the actual drawing from it. Strictly speaking, contexts don't have to synchronize among each other according to the specification, and I guess drawing from a VBO that was still being updated is something macOS doesn't like.

I also had to add `XInitThreads()` just to keep Xlib happy on Linux and FreeBSD.

**Comment by mantognini** [comment]
screen shot 2018-03-25 at 00 44 59


:/

**Comment by mantognini** [comment]
With those changes (and adding the appropriate linker setting), it works like a charm. 👍

File: examples/island/Island.cpp [hunk 3]
```diff
@@ -0,0 +1,540 @@
...
+        parametersText.setCharacterSize(14);
+        parametersText.setFillColor(sf::Color::White);
+        parametersText.setOutlineColor(sf::Color::Black);
+        parametersText.setOutlineThickness(2.0f);
+        parametersText.setPosition(5.0f, 5.0f);
+
+        if (!terrainShader.loadFromFile("resources/terrain.vert", "resources/terrain.frag"))
+            return EXIT_FAILURE;
+
+        sf::Image terrainImage;
```
[the remark this answers is not in the document]
Comment (mantognini) [reply examples/island/Island.cpp]: I'm curious too.

File: examples/island/Island.cpp [hunk 4]
```diff
@@ -0,0 +1,540 @@
...
+            }
+
+            // Arrow key pressed:
+            if (prerequisitesSupported && (event.type == sf::Event::KeyPressed))
+            {
+                switch (event.key.code)
+                {
+                    case sf::Keyboard::Down:  currentSetting = (currentSetting + 1) % settingCount; break;
+                    case sf::Keyboard::Up:    currentSetting = (currentSetting + settingCount - 1) % settingCount; break;
+                    case sf::Keyboard::Left:  *settings[currentSetting] -= 0.1f; generateTerrain(terrain); break;
```
[the remark this answers is not in the document]
Comment (mantognini) [reply examples/island/Island.cpp]: I wouldn't go that for for this example because it is not related to vertex buffer and cie.

File: examples/island/Island.cpp [hunk 5]
```diff
@@ -0,0 +1,540 @@
...
+
+        if (prerequisitesSupported)
+        {
+            terrainShader.setUniform("lightFactor", lightFactor);
+            window.draw(terrain, terrainStates);
+
+            osstr.str("");
+            osstr << "Frame: " << clock.restart().asMilliseconds() << "ms\n"
+                  << "perlinOctaves: " << perlinOctaves << "\n"
+                  << (settings[currentSetting] == &perlinFrequency ? "> " : "") << "perlinFrequency: " << perlinFrequency << "\n"
```
[the remark this answers is not in the document]
Comment (mantognini) [reply examples/island/Island.cpp]: +1 for the comment, the rest I don't mind.

**Comment by mantognini** [comment]
Github seems to display my comments out of context. See @MarioLiebisch's review for better understanding. https://github.com/SFML/SFML/pull/1352#pullrequestreview-92022269

**Comment by binary1248** [comment]
@SFML/sfml Any suggestions on how to enable to user to explicitly flush/finish an `sf::Context` without having to call `glFlush()`/`glFinish()` in their own code? I was thinking about adding something like `sf::Context::flush()` and `sf::Context::finish()`.

It is obvious from situations like the one we have here that more advanced use cases might demand more explicit control of the OpenGL command stream. In newer OpenGL, one would have access to `glFenceSync()`, however we have to emulate it using the above mentioned functions instead.

**Comment by LaurentGomila** [comment]
What's wrong with calling `glFlush(`) / `glFinish()` directly?

**Comment by binary1248** [comment]
As seen in the discussion above, this requires the user to not only `#include <SFML/OpenGL.hpp>` but to also link in OpenGL themselves even if they are linking dynamically. I don't have the feeling that this one function call justifies having to go that deep. The "promise" if you will is that as long as you limit yourself to sfml-graphics you wouldn't have to deal directly with OpenGL yourself. We kind of break this promise here.

**Comment by LaurentGomila** [comment]
But why would you have to call these GL functions if you're not making any other GL call? If this is needed for internal operation of SFML, why can't we call them appropriately?

**Comment by binary1248** [comment]
Because calling them when unnecessary (which is probably 99% of the time) will lead to severe performance penalties.

As an example: you would never call `glFinish()` anywhere where you do your rendering. It causes the thread to block until the GPU command queue is completely empty. Because this example is meant to demonstrate a very very special use case in which we are only transferring data from secondary threads, it is acceptable to call `glFinish()` here.

Like I said, this isn't something we can/should assume anything about. Most of the time, it works without it, but as @mantognini will be able to tell you, `glFinish()` has been the only way to make macOS happy because it has a really picky (albeit still specification compliant) OpenGL implementation.

**Comment by eXpl0it3r** [comment]
Should we merge this example, discuss potential API changes on the forum and adapt the code afterwards if needed.

**Comment by binary1248** [comment]
I can adjust it to `#include <SFML/OpenGL.hpp>` and call `glFinish()` but then it would lose its status as a "sfml-graphics only example".

**Comment by LaurentGomila** [comment]
> a very very special use case

Maybe *too* special? Is it worth modifying the public API for it? As far as I know, nobody ever had issues with such a code (probably because nobody ever wrote such a code 😬), nor asked for a wrapped glFinish() / glFlush().

**Comment by binary1248** [comment]
Before `sf::VertexBuffer` was available, our general advice to people regarding multi-threading and OpenGL was generally "don't do it... it won't give you the results you expect". This is still true if you limit OpenGL usage to pure drawing.

When you factor in VBO data uploads now, it changes the story a bit. VBO data uploads are (when you plan your application right) more or less order independent. It doesn't matter when exactly they happen if you can get your data dependency management right. This allows them to be offloaded to threads better than draw operations.

This is also the general direction that is given by Khronos when it comes to AZDO or "modern OpenGL usage" and also the reason why Vulkan is so multi-core friendly by design.

In short: `sf::VertexBuffer` enables new use cases, including this one, and we should support users who want to take advantage of this capability.

**Comment by LaurentGomila** [comment]
Ok, I see. Then what about collecting some feedback & usage of `sf::VertexBuffer` before deciding API modifications?

**Comment by mantognini** [comment]
This example is already interesting as is, so I wouldn't mind it being explicitly depending on OGL and not just graphics. We can improve it later when/if we change the API. It also serves as a good example showing the kind of 'problem' one would have to deal with. (Although, should SFML be capable of handling that on its own is also desirable.) 

There was also this discussion of multiple backend. I don't know where it was left of (out of scope?) but there might be a more general concept than `flush`, maybe?

**Comment by binary1248** [comment]
In Direct3D there is `ID3D11DeviceContext::Flush` and in Vulkan... well... since everything is so explicit there, you can't really get into the situation in which an explicit flush would be necessary.

Background information: `glFlush` is only necessary because for some strange reason, an OpenGL implementation is allowed to withhold commands in its CPU-side queue for an indefinite amount of time as long as a user does not call any OpenGL functions. It only has to queue CPU-side in the first place if it determines that the GPU-side queue is full. This means that buffer update operations in secondary contexts that don't immediately make it to the GPU might never be completed without a manual flush every now and then. Strange I know... but that is the way it is. Whether this is also the exact reasoning for `Flush` in Direct3D I cannot say.

Back to the matter at hand...

I realize that perhaps we haven't provided enough fine-grained control to the user to reliably make use of the functionality I mentioned in my previous comment. Further extensions would have to be made to the API in order to properly support this. Furthermore, since the majority of the time spent now is actually re-computing the terrain, offloading the actual buffer updates to secondary contexts as well does make a difference, however it might be so small that it isn't worth the extra effort required to make it work on all platforms. As such I will adjust the example to move the buffer operations into the main thread and push a new version. We can talk about optimizing buffer usage at some other time.

**Comment by binary1248** [comment]
All the issues should be fixed now. Also made it a bit more user friendly.

File: examples/island/Island.cpp [hunk 6]
```diff
@@ -0,0 +1,579 @@
...
+        statusText.setOutlineColor(sf::Color::Black);
+        statusText.setOutlineThickness(2.0f);
+
+        hudText.setCharacterSize(14);
+        hudText.setFillColor(sf::Color::White);
+        hudText.setOutlineColor(sf::Color::Black);
+        hudText.setOutlineThickness(2.0f);
+        hudText.setPosition(5.0f, 5.0f);
+
+        if (!terrainShader.loadFromFile("resources/terrain.vert", "resources/terrain.frag"))
```
Comment (mantognini) [inline examples/island/Island.cpp]: I'd move this to line 92 so that when calling `generateTerrain` the prerequisites are always met.

**Comment by binary1248** [comment]
@mantognini Fixed.
